### PR TITLE
[RLlib] Migrate maml, mbmpo, marwil to use sub-classing implementation.

### PR DESCRIPTION
**Pull request by gjoliver**
## Why are these changes needed?

Migrate MAML, MBMPO, and MARWIL to use the new sub-classing implementation.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [*] Unit tests
   - [*] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/policy/policy.py [hunk 1]
```diff
@@ -155,6 +157,20 @@ def __init__(
...
 
+        # Initialize view requirements.
+        self.init_view_requirements()
+
+        # Whether the Model's initial state (method) has been added
+        # automatically based on the given view requirements of the model.
+        self._model_init_state_automatically_added = False
+
+    @DeveloperAPI
+    def init_view_requirements(self):
```
Comment (sven1977) [inline rllib/policy/policy.py:168]: Nice!

File: rllib/algorithms/mbmpo/mbmpo_torch_policy.py [hunk 2]
```diff
@@ -1,135 +1,90 @@
...
+                + "Consider reshaping this into a single dimension Box space "
+                "or using the multi-agent API."
+            )
+
+        config = dict(ray.rllib.algorithms.mbmpo.mbmpo.DEFAULT_CONFIG, **config)
+        super().__init__(observation_space, action_space, config)
+
+    def make_model_and_action_dist(
+        self,
+    ) -> Tuple[ModelV2, Type[TorchDistributionWrapper]]:
```
Comment (sven1977) [inline rllib/algorithms/mbmpo/mbmpo_torch_policy.py:41]: Nite: Don't have to do this in this PR: We should unify the APIs here for torch: `make_model_and_action_dist` and tf: `make_model` (action_dist?? :) ).

**Review by sven1977 — APPROVED** [review]
Looks great, just a few questions and one small nit.

File: rllib/algorithms/maml/maml_tf_policy.py [hunk 3]
```diff
@@ -415,41 +352,154 @@ def update_kls(self, sampled_kls):
...
+                    clip_param=self.config["clip_param"],
+                    vf_clip_param=self.config["vf_clip_param"],
+                    vf_loss_coeff=self.config["vf_loss_coeff"],
+                    use_gae=self.config["use_gae"],
+                )
+
+            return self.loss_obj.loss
+
+        @override(base)
+        def optimizer(
```
[the remark this answers is not in the document]
Comment (gjoliver) [reply rllib/algorithms/maml/maml_tf_policy.py:470]: yeah, probably, ok adding this to post migration cleanup list?

File: rllib/policy/tf_policy.py [hunk 4]
```diff
@@ -784,6 +784,15 @@ def extra_compute_action_feed_dict(self) -> Dict[TensorType, TensorType]:
 
     @DeveloperAPI
     def extra_compute_action_fetches(self) -> Dict[str, TensorType]:
+        # It is pretty important to cache this, since action_fetches are
+        # called every single iteration. Things will be really slow if
+        # we try to compute the fetches live every time.
```
[the remark this answers is not in the document]
Comment (gjoliver) [reply rllib/policy/tf_policy.py:789]: yep, this gets called for every action computation call.
this is TF1, so everything is statically constructed, and you know which tensors you need to fetch beforehand.
I think mostly what this function does is to say I need the value function fetch for this particular algorithm, and then we cache the fact that we need to fetch the vf tensor.

File: rllib/algorithms/mbmpo/mbmpo_torch_policy.py (hunk 2 above)
Comment (gjoliver) [reply rllib/algorithms/mbmpo/mbmpo_torch_policy.py:41]: definitely, we should.

File: rllib/policy/tf_policy.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/policy/tf_policy.py]: Oh, got it. Makes sense, so we kept adding to the graph :( 
Thanks for the explanation.

File: rllib/algorithms/maml/maml_tf_policy.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/maml/maml_tf_policy.py:470]: Ok, sounds good!

**Pull request merged**